File: CSharp/Shapes/Sphere.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CSharp
{
    public class Sphere : RTObject
    {
        private BoundingBox _bounds;

        public Sphere()
        {
            _bounds = new BoundingBox(new Point(-1, -1, -1), new Point(1, 1, 1));
        }

        public static Sphere Glass
        {
            get
            {
                var sphere = new Sphere();
                sphere.Material.Transparency = 1.0;
                sphere.Material.RefractiveIndex = 1.5;

                return sphere;
            }
        }

        public override BoundingBox BoundsOf => _bounds;

        protected override Intersections LocalIntersect(Ray ray)
        {
            var result = new Intersections();
            var sphereToRay = new Vector(ray.Origin - Point.Zero);
            var a = ray.Direction.Dot(ray.Direction);
            var b = 2.0 * ray.Direction.Dot(sphereToRay);
            var c = sphereToRay.Dot(sphereToRay) - 1;
            var discriminant = b * b - 4 * a * c;

            if (discriminant < 0.0)
            {
                return result;
            }
            else
            {
                result.Add(new Intersection((-b - Math.Sqrt(discriminant)) / (2 * a), this));
                result.Add(new Intersection((-b + Math.Sqrt(discriminant)) / (2 * a), this));
            }

            return result;
        }

        protected override Vector LocalNormal(Point p, Intersection hit)
        {
            var point = p - Point.Zero;
            return new Vector(point.x, point.y, point.z);
        }

        public override bool Equals(object obj)
        {
            var other = obj as Sphere;
            return base.Equals(obj);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine("Sphere", Material, Transform);
        }

        public override string ToString()
        {
            return $"Sphere ({ID}): {Material} {Transform}";
        }
    }
}

File: Java/src/main/java/Shapes/Sphere.java
package RTC.Shapes;

import RTC.Math.*;
import RTC.Material.*;
import java.util.Objects;

public class Sphere extends Shape {
    public Sphere() {
        super();
    }

    public Vector localNormalAt(Point localPoint) {
        return new Vector(localPoint.sub(Point.Zero));
    }

    public Intersections localIntersect(Ray r) {
        Intersections result = new Intersections();
        Tuple sphereToRay = r.getOrigin().sub(Point.Zero);
        Vector dir = r.getDirection();
        double a = dir.dot(dir);
        double b = 2.0 * dir.dot(sphereToRay);
        double c = sphereToRay.dot(sphereToRay) - 1.0;

        double discriminant = (b * b) - 4.0 * a * c;

        if (discriminant >= 0.0) {
            result.add(new Intersection((-b - Math.sqrt(discriminant)) / (2 * a), this));
            result.add(new Intersection((-b + Math.sqrt(discriminant)) / (2 * a), this));
        }

        return result;
    }

    @Override
    public boolean equals(Object o)
    {
        if (o == this) {
            return true;
        }

        if ((o == null) || !(o instanceof Sphere)) {
            return false;
        }

        Sphere r = (Sphere) o;
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return String.format("Sphere Id: %d", id);
    }
}

